aptos-move/framework/src/natives/hash.rs

```diff
@@ -0,0 +1,59 @@
+use move_deps::{
+    move_binary_format::errors::PartialVMResult,
+    move_core_types::{
+        gas_schedule::{GasAlgebra, GasCost},
+        vm_status::sub_status::NFE_BCS_SERIALIZATION_FAILURE,
+    },
+    move_vm_runtime::native_functions::NativeContext,
+    move_vm_types::{
+        gas_schedule::NativeCostIndex,
+        loaded_data::runtime_types::Type,
+        natives::function::{native_gas, NativeResult},
+        pop_arg,
+        values::{values_impl::Reference, Value},
+    },
+};
+use smallvec::smallvec;
+use std::{collections::VecDeque, hash::Hasher};
+
+/// Serialize the MoveValue with BCS and then feed the bytes into SipHasher. This is not
+/// cryptographically secure.
+pub fn native_sip_hash(
+    context: &mut NativeContext,
+    mut ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.len() == 1);
+    debug_assert!(args.len() == 1);
+
+    let ref_to_val = pop_arg!(args, Reference);
+
+    let arg_type = ty_args.pop().unwrap();
+    // delegate to the BCS serialization for `Value`
+    let serialized_value_opt = match context.type_to_type_layout(&arg_type)? {
+        None => None,
+        Some(layout) => ref_to_val.read_ref()?.simple_serialize(&layout),
+    };
+    let serialized_value = match serialized_value_opt {
+        None => {
+            let cost = native_gas(context.cost_table(), NativeCostIndex::BCS_TO_BYTES, 1);
+            return Ok(NativeResult::err(cost, NFE_BCS_SERIALIZATION_FAILURE));
+        }
+        Some(serialized_value) => serialized_value,
+    };
+
+    // cost is proportional to the size of the serialized value
+    let bcs_cost = native_gas(
+        context.cost_table(),
+        NativeCostIndex::BCS_TO_BYTES,
+        serialized_value.len(),
+    );
+    // SipHash of the serialized bytes
+    let mut hasher = siphasher::sip::SipHasher::new();
```

Comment:
n00b Q: With the key0 and key1 known (to be 0), does SIP offer anything better than other commonly used hash functions?  I mean, why SIP?

Reply:
no idea, I just copied rust hashmap's default hasher

Reply:
@alinush Science, please. :trollface:

Reply:
@msmouse, I'm not sure I understand what you mean by `key0` and `key1` :) 

If the question to "be science'd" is why SIP, I am not sure why @zekun000 chose it over others. As far as I can tell, we want to pick the fastest hash function that gives us as uniform as possible outputs. (We don't need collision resistance here for security, although it would help for performance, assuming it's not too costly.)

@zekun000, why not [ahash](https://crates.io/crates/ahash), which apparently could be much faster (see their graphs)?

Reply:
it's just the default hasher for the rust hashmap, happy to change it to anything else

Reply:
@alinush this:
https://docs.rs/siphasher/latest/src/siphasher/sip.rs.html#149
I think the point of using SipHasher in rust and other languages is the keys are decided randomly at runtime, so it's hard to create artificial collisions. https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#2939-2960

For a blockchain that doesn't matter, hence the ask..

Reply:
I see. So, is the question_"Does SIP offer anything better than other commonly used hash functions **when initialized with key0 and key1 to zero**?"_

Reply:
@alinush  yes, I think we just need a fast and even hash function here.

Reply:
We could swap in different hash functions then and benchmark the execution of various smart contracts that use hash maps.